src/libraries/Common/src/SourceGenerators/TypeModelHelpers.cs

```diff
@@ -0,0 +1,102 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using System.Collections.Generic;
+
+namespace SourceGenerators
+{
+    internal static class TypeModelHelpers
+    {
+        private static readonly SymbolDisplayFormat s_minimalDisplayFormat = new SymbolDisplayFormat(
+            globalNamespaceStyle: SymbolDisplayGlobalNamespaceStyle.Omitted,
+            typeQualificationStyle: SymbolDisplayTypeQualificationStyle.NameAndContainingTypes,
+            genericsOptions: SymbolDisplayGenericsOptions.IncludeTypeParameters,
+            miscellaneousOptions: SymbolDisplayMiscellaneousOptions.UseSpecialTypes);
+
+        public static string ToIdentifierCompatibleSubstring(this ITypeSymbol type, bool useUniqueName = false)
+        {
+            if (type is IArrayTypeSymbol arrayType)
+            {
+                int rank = arrayType.Rank;
+                string suffix = rank == 1 ? "Array" : $"Array{rank}D"; // Array, Array2D, Array3D, ...
+                return ToIdentifierCompatibleSubstring(arrayType.ElementType) + suffix;
+            }
+
+            string name = GetName(type);
+
+            if (type is not INamedTypeSymbol namedType || !namedType.IsGenericType)
+            {
+                return name;
+            }
+
+            StringBuilder sb = new();
+
+            sb.Append(name);
+
+            foreach (ITypeSymbol genericArg in namedType.GetAllTypeArgumentsInScope())
+            {
+                sb.Append(ToIdentifierCompatibleSubstring(genericArg));
+            }
+
+            return sb.ToString();
+
+            string GetName(ITypeSymbol type) => useUniqueName
+                ? ToIdentifierCompatibleSubstring(type.ToMinimalDisplayString())
```

Comment:
>ToIdentifierCompatibleSubstring [](http://example.com/codeflow?start=18&length=31)

You can pass the created `sb` here instead of creating another StringBuilder object inside this method.

Reply:
Will switch thanks. There was a reason I didn't do that in an earlier iteration but it no longer holds.